compiler/resolution/functionResolution.cpp

```diff
@@ -4931,14 +4931,56 @@ createFunctionAsValue(CallExpr *call) {
 
   FnSymbol* captured_fn = visibleFns.head();
 
-  //Check to see if we've already cached the capture somewhere
-  if (functionCaptureMap.find(captured_fn) != functionCaptureMap.end()) {
-    return new CallExpr(functionCaptureMap[captured_fn]);
+  //
+  // Is this a case where we need a C function pointer or a Chapel
+  // first-class function / closure?
+  //
+  bool cFunctionPtr = false;
+
+  //
+  // If the capture call has two arguments, it implies that we added a
+  // sentinel in scopeResolve.cpp because we detected a call to
+  // c_fnPtrTo()).  So this should just be a simple function pointer,
+  // not a Chapel first-class function.
+  //
+  if (call->numActuals() == 2) {
```

Comment:
This isn't the way of determining whether we're in a capture call, but the way of distinguishing whether we're capturing the function for C or for Chapel.  I agree it's subtle, but AFAIK, primitives don't have a way of creating named arguments (which seemed like a small step to improve it).  I'm open to suggestions here, but didn't come up with any particularly attractive solutions myself.